toolchain/lex/tokenized_buffer.cpp

```diff
@@ -352,32 +376,92 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
     CARBON_DCHECK(source_text.substr(position).startswith("//"));
 
     // Any comment must be the only non-whitespace on the line.
-    if (set_indent_) {
+    const auto* line_info = current_line_info();
+    if (LLVM_UNLIKELY(position != line_info->start + line_info->indent)) {
       CARBON_DIAGNOSTIC(TrailingComment, Error,
                         "Trailing comments are not permitted.");
 
       emitter_.Emit(source_text.begin() + position, TrailingComment);
     }
 
     // The introducer '//' must be followed by whitespace or EOF.
-    if (position + 2 < static_cast<ssize_t>(source_text.size()) &&
+    if (LLVM_UNLIKELY(position + 2 <
+                      static_cast<ssize_t>(source_text.size())) &&
         !IsSpace(source_text[position + 2])) {
```

Comment:
This expectation looks reversed to me -- I'd expect the `//` is almost always not at the very end of the file. Should this perhaps be either:
```suggestion
    if (LLVM_LIKELY(position + 2 <
                    static_cast<ssize_t>(source_text.size())) &&
        !IsSpace(source_text[position + 2])) {
```
or
```suggestion
    if (LLVM_UNLIKELY(position + 2 <
                          static_cast<ssize_t>(source_text.size()) &&
                      !IsSpace(source_text[position + 2]))) {
```
? If it's correct as-is, a comment explaining why would be helpful :)

Reply:
I meant for it to be on the entire condition (your second suggestion), but I'd rather it actually only apply to the is-space, so I'll do that instead.